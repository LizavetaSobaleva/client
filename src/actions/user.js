import axios from "axios"
import { setUser } from '../reducers/userReducer'


export const instanceAxios = axios.create({
     baseURL: process.env.REACT_APP_API_URL + "/api/"
})

export const registration = async (name, email, password, dispatch) => {
   try {
        const response = await instanceAxios.post('auth/registration', {
            name,
            email,
            password
        })
        alert(response.data.message)
        dispatch(login(email, password))
   } catch (e) {
        alert(e.response.data.message)
   }
}

export const login = (email, password) => {
     return async dispatch => {
          try {
               const response = await instanceAxios.post('auth/login', {
                   email,
                   password
               })
               dispatch(setUser(response.data.user))
               localStorage.setItem('token', response.data.token)
          } catch (e) {
               alert(e.response.data.message)
          }
     }
}

export const auth = () => {
     return async dispatch => {
          try {
               const response = await instanceAxios.get('auth/auth',
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               localStorage.setItem('token', response.data.token)
               dispatch(setUser(response.data.user))
          } catch (e) {
               console.log(e.response.data.message)
               localStorage.removeItem('token')
          }
     }
}

export const uploadAvatar = (file) => {
     return async dispatch => {
          try {
               const formData = new FormData()
               formData.append('file', file)
               const response = await instanceAxios.post('files/avatar', formData,
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               dispatch(setUser(response.data))
          } catch (e) {
               console.log(e)
          }
     }
}

export const deleteAvatar = () => {
     return async dispatch => {
          try {
               const response = await instanceAxios.delete('files/avatar',
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               dispatch(setUser(response.data))
          } catch (e) {
               console.log(e)
          }
     }
}