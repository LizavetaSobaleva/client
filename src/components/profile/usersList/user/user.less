@import '../../../../utils/colors.less';

@import '../../../../utils/colors.less';

.user {
    padding: 8px 12px;
    border-bottom: 1px solid @lightgrey;
    display: grid;
    grid-template-columns: 1.5fr 2fr 2fr 2fr 0.5fr;
    color: @subtext;
    font-weight: 400;
    align-items: center;
    transition: ease-in 50ms all;

    &:hover {
        background-color: #f9fafc;
    }

    &__name {
        grid-column-start: 1;
        width: 98%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__email {
        grid-column-start: 2;
        width: 98%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__status {
        grid-column-start: 3;
        justify-self: center;

        select {
            padding: 4px 16px;
            border: none;
            border-radius: 16px;
            background-color: #f9fafc;
            color: @primary;
            font-size: 16px;
            cursor: pointer;
        
            &:focus {
              outline: none;
            }
          }
    }

    &__size {
        grid-column-start: 4;
        width: auto;
        display: flex;
        flex-direction: column;
        font-size: 12px;
        padding: 4px 20px;

        .progressBar {
            width: auto;
            padding: 4px 0;

            &__line, &__upload {
                height: 8px;
            }

            &__percent {
                display: none;
            }
        }
    }

    &__actions {
        grid-column-start: 5;

        .primaryButton {
            padding: 8px 24px;
        }
    }
}

@media only screen and (max-width: 660px){
    .user {
        grid-template-columns: 1.5fr 2fr 2fr;

        &__name {
            margin-left: 4px;
        }

        &__email {
            display: none
        }

        &__status {
            grid-column-start: 2;
        }
    
        &__size {
            grid-column-start: 3;
        }
    }
}