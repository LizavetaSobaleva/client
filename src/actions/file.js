import { hideLoader, showLoader } from "../reducers/appReducer"
import { addFile, deleteFileAction, setFiles, setStructure } from "../reducers/fileReducer"
import { addUploadFile, changeUploadFile, showUploader } from "../reducers/uploadReducer"
import { instanceAxios } from "./user"

export function getFiles(dirId) {
  return async (dispatch) => {
    try {
      dispatch(showLoader())
      const response = await instanceAxios.get(`files${dirId ? '?parent='+dirId : ''}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setFiles(response.data))
    } catch (e) {
      console.log(e.response.data.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}

export function createDir(dirId, name) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.post(`files`, {
          name,
          parent: dirId,
          type: 'dir'
        },
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
        )
      dispatch(addFile(response.data))
    } catch (e) {
      alert(e.response.data.message)
    }
  }
} 

export function uploadFile(file, dirId) {
  return async (dispatch) => {
    try {
      const formData = new FormData()
      formData.append('file', file)
      if(dirId) {
        formData.append('parent', dirId)
      }
      const uploadFile = {name: file.name, process: 0, id: Date.now()}
      dispatch(showUploader())
      dispatch(addUploadFile(uploadFile))
      const response = await instanceAxios.post(`files/upload`, formData, {
                headers: {Authorization: `Bearer ${localStorage.getItem('token')}`},
                onUploadProgress: progressEvent => {
            const totalLength = progressEvent.event.lengthComputable ? progressEvent.event.total : progressEvent.event.target.getResponseHeader('content-length') || progressEvent.event.target.getResponseHeader('x-decompressed-content-length')
            if(totalLength) {
              uploadFile.progress = Math.round((progressEvent.loaded * 100) / totalLength)
              dispatch(changeUploadFile(uploadFile))
            }
          }
        }
      )
      dispatch(addFile(response.data))
    } catch (e) {
      alert(e.response.data.message)
    }
  }
} 

export async function downloadFile(file) {
  try {
    const response = await instanceAxios.get(`files/download?id=${file._id}`, {
      responseType: 'blob',
      headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
    )
    const blob = new Blob([response.data], { type: response.headers['content-type']})

    const downloadUrl = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = downloadUrl
    link.download = file.name

    document.body.appendChild(link)
    link.click()
    link.remove()
  } catch (e) {
    console.log(e.response.data.message)
  }
}

export function deleteFile(file) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.delete(`files?id=${file._id}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(deleteFileAction(file._id))
      alert(response.data.message)
    } catch (e) {
      alert(e.response.data.message)
    }
  }
}

export function searchFiles(search) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.get(`files/search?search=${search}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setFiles(response.data))
    } catch (e) {
      alert(e?.response?.data?.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}

export function getStructure() {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.get(`files/structure`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setStructure(response.data))
    } catch (e) {
      alert(e?.response?.data?.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}