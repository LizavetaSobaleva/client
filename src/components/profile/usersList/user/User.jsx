import React, { useState } from "react";
import "./user.less";
import sizeFormat from "../../../../utils/sizeFormat";
import ProgressBar from "../../../UI/progressBar/ProgressBar";
import { useDispatch } from "react-redux";
import { changeUserStatus } from "../../../../actions/user";
import PrimaryButton from "../../../UI/primaryButton/PrimaryButton";
import SecondaryButton from "../../../UI/secondaryButton/SecondaryButton";
import Label from "../../../UI/label/Label";
import Selector from "../../../UI/selector/Selector";

const User = ({ user, fetchUsers }) => {
  const dispatch = useDispatch();
  const usedSpace = ((user.usedSpace / user.diskSpace) * 100).toFixed();
  const [status, setStatus] = useState(user.status);
  const [isEditing, setIsEditing] = useState(false);

  const handleStatusChange = (newStatus) => {
    setStatus(newStatus);
  };

  const handleSave = async () => {
    await dispatch(changeUserStatus(user._id, status));
    setIsEditing(false);
    fetchUsers();
  };

  const statusOptions = [
    { value: 'standard', label: 'standard' },
    { value: 'premium', label: 'premium' },
    { value: 'admin', label: 'admin' },
  ];

  return (
    <div className="user">
      <div className="user__name">{user.name}</div>
      <div className="user__email">{user.email}</div>

      <div className="user__status">
        {isEditing ? (
          <Selector value={status} options={statusOptions} onChange={handleStatusChange} />
        ) : (
          <Label status={user.status}>{user.status}</Label>
        )}
      </div>

      <div className="user__size">
        {sizeFormat(user.usedSpace)} from {sizeFormat(user.diskSpace)}
        <ProgressBar progress={usedSpace} />
      </div>

      <div className="user__actions">
        {isEditing ? (
          <PrimaryButton onClick={handleSave}>Save</PrimaryButton>
        ) : (
          <SecondaryButton onClick={() => setIsEditing(true)}>Edit</SecondaryButton>
        )}
      </div>
    </div>
  );
};

export default User;