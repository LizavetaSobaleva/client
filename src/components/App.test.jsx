import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import App from './App';
import { auth } from '../actions/user';

jest.mock('../actions/user', () => ({
  auth: jest.fn().mockReturnValue({ type: 'SET_USER' }),
}));

jest.mock('../actions/file', () => ({
    getFiles: jest.fn(() => ({ type: 'GET_FILES' })),
    getStructure: jest.fn(() => ({ type: 'GET_STRUCTURE' })),
}));

const mockStore = configureStore([]);
let store;

const initialState = {
    user: { isAuth: false, currentUser: { name: 'John Doe', avatar: 'avatar.png' } },
    files: { currentDir: 'root', dirStructure: [], files: [] },
    app: { view: 'list' },
    upload: { files: [], isVisible: false },
  };

const renderComponent = (store) =>
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );

describe('App Component', () => {
  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  test('renders App without crashing', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('app')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders login container when not authenticated', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('loginContainer')).toBeInTheDocument();
    expect(screen.queryByTestId('contentContainer')).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders content container when authenticated', () => {
    store = mockStore({
      ...initialState,
      user: { isAuth: true, currentUser: { name: 'John Doe', avatar: 'avatar.png' } },
    });
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('contentContainer')).toBeInTheDocument();
    expect(screen.queryByTestId('loginContainer')).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('auth action is dispatched on mount', () => {
    renderComponent(store);
    expect(store.dispatch).toHaveBeenCalledWith(auth());
  });
});
