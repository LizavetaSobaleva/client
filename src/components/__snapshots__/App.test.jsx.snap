// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders App without crashing 1`] = `
<DocumentFragment>
  <div
    class="app"
    data-testid="app"
  >
    <div
      class="login"
      data-testid="loginContainer"
    >
      <div
        class="navbar"
        data-testid="navbar"
      >
        <div
          class="navbar__content"
          data-testid="navbarContent"
        >
          <div
            class="navbar__header"
            data-testid="navbarHeader"
          >
            <img
              alt=""
              class="navbar__logo"
              data-testid="navbarLogo"
              src="test-file-stub"
            />
            <span>
              MERN cloud
            </span>
          </div>
          <div
            class="navbar__btns"
            data-testid="navbarButtons"
          >
            <div
              class="navbar__login"
              data-testid="loginButton"
            >
              <a
                aria-current="page"
                class="active"
                href="/login"
              >
                Log in
              </a>
            </div>
            <div
              class="navbar__registration"
              data-testid="registerButton"
            >
              <a
                class=""
                href="/registration"
              >
                Sign up
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="authorization"
      >
        <div
          class="authorization__container"
        >
          <div
            class="authorization__label"
            data-testid="formTitle"
          >
            <h1>
              Log in to cloud
            </h1>
          </div>
          <div
            class="authorization__form"
          >
            <div
              class="authorization__item"
            >
              <label
                for="emailInput"
              >
                Email
              </label>
              <input
                data-testid="emailInput"
                id="emailInput"
                placeholder="Enter your email"
                type="email"
                value=""
              />
            </div>
            <div
              class="authorization__item"
            >
              <label
                for="passwordInput"
              >
                Password
              </label>
              <input
                data-testid="passwordInput"
                id="passwordInput"
                placeholder="Enter your password"
                type="password"
                value=""
              />
            </div>
            <button
              class="primaryButton"
              data-testid="loginButton"
            >
              Log in
            </button>
          </div>
          <div
            class="authorization__login"
            data-testid="registrationLink"
          >
            <a
              href="/registration"
            >
              Create an account
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App Component renders content container when authenticated 1`] = `
<DocumentFragment>
  <div
    class="app"
    data-testid="app"
  >
    <div
      class="sidebar "
      data-testid="sidebar"
    >
      <div
        class="sidebar__content"
        data-testid="sidebarContent"
      >
        <a
          aria-current="page"
          class="active"
          data-testid="homeLink"
          href="/"
        >
          <div
            class="sidebar__header"
            data-testid="sidebarHeader"
          >
            <img
              alt=""
              class="sidebar__logo"
              data-testid="sidebarLogo"
              src="test-file-stub"
            />
            <div
              class="sidebar__title"
            >
              MERN cloud
            </div>
          </div>
        </a>
        <div
          class="sidebar__navigation"
          data-testid="sidebarNavigation"
        >
          <div
            class="navigation"
            data-testid="navigation"
          >
            <div
              class="navigation__item"
              data-testid="navigationItem"
            >
              <a
                aria-current="page"
                class="active"
                data-testid="navLink"
                href="/"
              >
                <div
                  class="navigation__title"
                  data-testid="navigationTitle"
                >
                  <img
                    alt=""
                    class="navigation__icon"
                    data-testid="navigationIcon"
                    src="test-file-stub"
                  />
                  <div
                    class="navigation__name"
                  >
                    My Cloud
                  </div>
                </div>
              </a>
              <div
                class="navigation__content visible"
                data-testid="navigationContent"
              />
            </div>
          </div>
        </div>
        <div
          class="sidebar__logout"
          data-testid="sidebarLogout"
        >
          <button
            class="secondaryButton"
            data-testid="logoutButton"
          >
            <img
              alt=""
              class="sidebar__logout-icon"
              data-testid="logoutIcon"
              src="test-file-stub"
            />
            <div
              class="sidebar__logout-text"
            >
              Log out
            </div>
          </button>
        </div>
      </div>
      <div
        class="sidebar__compressBtn"
        data-testid="compressButton"
      >
        <div
          class="dropdown"
          data-testid="dropdownButton"
        >
          <img
            alt=""
            class="dropdown__icon "
            data-testid="dropdownIcon"
            src="test-file-stub"
          />
        </div>
      </div>
    </div>
    <div
      class="content"
      data-testid="contentContainer"
    >
      <div
        class="navbar"
        data-testid="navbar"
      >
        <div
          class="navbar__content"
          data-testid="navbarContent"
        >
          <div
            class="navbar__header"
            data-testid="searchHeader"
          >
            <input
              data-testid="searchInput"
              placeholder="Type to search..."
              type="search"
              value=""
            />
          </div>
          <div
            class="navbar__user"
            data-testid="navbarUser"
          >
            <div
              class="navbar__message"
              data-testid="userMessage"
            >
              Hello, John Doe!
            </div>
            <a
              class=""
              href="/profile"
            >
              <img
                alt=""
                class="navbar__avatar"
                data-testid="userAvatar"
                src="undefined/avatar.png"
              />
            </a>
          </div>
        </div>
      </div>
      <div
        class="disk"
      >
        <div
          class="disk__btns"
        >
          <button
            class="primaryButton"
            data-testid="diskCreateBtn"
          >
            Create folder
          </button>
          <div
            class="disk__view"
          >
            <button
              class="secondaryButton"
              data-testid="listViewBtn"
            >
              <img
                alt=""
                class="disk__list"
                src="test-file-stub"
              />
            </button>
            <button
              class="secondaryButton"
              data-testid="plateViewBtn"
            >
              <img
                alt=""
                class="disk__plate"
                src="test-file-stub"
              />
            </button>
          </div>
        </div>
        <div>
          <div
            class="upload"
            data-testid="uploadArea"
          >
            <label
              class="upload__label"
              for="upload-input"
            >
              <img
                alt=""
                class="upload__img"
                src="test-file-stub"
              />
              <div
                class="upload__message"
                data-testid="uploadMessage"
              >
                Click and choose or drag files to this area to upload
              </div>
            </label>
            <input
              class="upload__input"
              id="upload-input"
              multiple=""
              type="file"
            />
          </div>
        </div>
        <div
          class="filelist"
        >
          <div
            class="filelist__header"
          >
            <div
              class="filelist__type"
            >
              Type
            </div>
            <div
              class="filelist__name"
            >
              Name
            </div>
            <div
              class="filelist__date"
            >
              Date
            </div>
            <div
              class="filelist__size"
            >
              File size
            </div>
            <div
              class="filelist__actions"
            />
          </div>
          <div
            class="filelist__back"
            data-testid="backBtn"
          >
            <img
              alt=""
              class="filelist__img"
              src="test-file-stub"
            />
            Back
          </div>
          <div />
        </div>
        <div
          class="filelist__empty"
        >
          Files not found
        </div>
        <div
          class="createdir"
          data-testid="testOutsideContent"
        >
          <div
            class="createdir__content"
            data-testid="testContent"
          >
            <div
              class="popup"
              data-testid="popup"
            >
              <div
                class="popup__header"
              >
                <div
                  class="popup__title"
                  data-testid="popupTitle"
                >
                  Create new folder
                </div>
                <button
                  class="secondaryButton"
                  data-testid="popupCloseBtn"
                >
                  <img
                    alt=""
                    class="popup__close"
                    data-testid="popupCloseIcon"
                    src="test-file-stub"
                  />
                </button>
              </div>
              <input
                data-testid="createInput"
                placeholder="Enter folder name..."
                type="text"
                value=""
              />
              <button
                class="primaryButton"
                data-testid="createBtn"
              >
                Create
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App Component renders login container when not authenticated 1`] = `
<DocumentFragment>
  <div
    class="app"
    data-testid="app"
  >
    <div
      class="login"
      data-testid="loginContainer"
    >
      <div
        class="navbar"
        data-testid="navbar"
      >
        <div
          class="navbar__content"
          data-testid="navbarContent"
        >
          <div
            class="navbar__header"
            data-testid="navbarHeader"
          >
            <img
              alt=""
              class="navbar__logo"
              data-testid="navbarLogo"
              src="test-file-stub"
            />
            <span>
              MERN cloud
            </span>
          </div>
          <div
            class="navbar__btns"
            data-testid="navbarButtons"
          >
            <div
              class="navbar__login"
              data-testid="loginButton"
            >
              <a
                aria-current="page"
                class="active"
                href="/login"
              >
                Log in
              </a>
            </div>
            <div
              class="navbar__registration"
              data-testid="registerButton"
            >
              <a
                class=""
                href="/registration"
              >
                Sign up
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="authorization"
      >
        <div
          class="authorization__container"
        >
          <div
            class="authorization__label"
            data-testid="formTitle"
          >
            <h1>
              Log in to cloud
            </h1>
          </div>
          <div
            class="authorization__form"
          >
            <div
              class="authorization__item"
            >
              <label
                for="emailInput"
              >
                Email
              </label>
              <input
                data-testid="emailInput"
                id="emailInput"
                placeholder="Enter your email"
                type="email"
                value=""
              />
            </div>
            <div
              class="authorization__item"
            >
              <label
                for="passwordInput"
              >
                Password
              </label>
              <input
                data-testid="passwordInput"
                id="passwordInput"
                placeholder="Enter your password"
                type="password"
                value=""
              />
            </div>
            <button
              class="primaryButton"
              data-testid="loginButton"
            >
              Log in
            </button>
          </div>
          <div
            class="authorization__login"
            data-testid="registrationLink"
          >
            <a
              href="/registration"
            >
              Create an account
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
